### ISSUE TITLE
Range Index Error while inserting job to jobs database

### ISSUE THREAD

**Issue by venuiyengar** [marklogic/marklogic-data-hub#3405]
I see the following error while running flows from quickstart in error log.

2019-12-13 16:04:51.733 Debug: IndexerEnv::putRangeIndex: XDMP-RANGEINDEX: xdmp:eval("xdmp.documentInsert(docUri, content, {permissions: xdmp.defaultP...", {content:{job:{jobId:"d93040df-3718-4fe2-9605-1e19d5acc0b2", flow:"pdd-to-patient-event", user:"admin", ...}}, docUri:"/jobs/d93040df-3718-4fe2-9605-1e19d5acc0b2.json", permissions:"xdmp.defaultPermissions().concat([xdmp.permission('flow-develope...", ...}, {database:"13123174156593553010", commit:"auto", update:"true", ...}) – Range index error: dateTime fn:doc('/jobs/d93040df-3718-4fe2-9605-1e19d5acc0b2.json')/job/text('timeEnded'): XDMP-LEXVAL: Invalid lexical value "N/A"
2019-12-13 16:04:51.812 Debug: IndexerEnv::putRangeIndex: XDMP-RANGEINDEX: xdmp:eval("xdmp.documentInsert(docUri, content, {permissions: xdmp.defaultP...", {content:{batch:{jobId:"d93040df-3718-4fe2-9605-1e19d5acc0b2", batchId:"44845808-7d8e-4c29-b9c3-cdbb1126862a", step:{name:"patient-mastering", description:"Step to master Patient entities", options:
{stepUpdate:true, additionalCollections:[], sourceQuery:"cts.collectionQuery(\"new-patient\")", .
, ...}, ...}}, docUri:"/jobs/batches/44845808-7d8e-4c29-b9c3-cdbb1126862a.json", permissions:"xdmp.defaultPermissions().concat([xdmp.permission('flow-develope...", ...}, {database:"13123174156593553010", commit:"auto", update:"true", ...}) – Range index error: dateTime fn:doc('/jobs/batches/44845808-7d8e-4c29-b9c3-cdbb1126862a.json')/batch/text('timeEnded'): XDMP-LEXVAL: Invalid lexical value "N/A"

I saw the similar error message while running quickStartTutorial while running sunrise tutorial

**Comment by xnikhil08** [comment]
@venuiyengar - Looks like you are getting 'N/A' as the value of the dateTime field 'timeEnded', which might be causing this issue. Could you please make sure the date fields in your database have accurate data. Also, please let me know the steps you have followed, in order for me to reproduce this issue at my end.

**Comment by venuiyengar** [comment]
@xnikhil08 
Please run this tutorial and you will see the error in errorlog.txt

https://developer.marklogic.com/learn/data-hub-quickstart/


2019-12-16 20:15:18.981 Debug: IndexerEnv::putRangeIndex: XDMP-RANGEINDEX: xdmp:eval("xdmp.documentInsert(docUri, content, {permissions: xdmp.defaultP...", {content:{job:{jobId:"cf09fd76-b7ad-4273-9428-f9068b374b0a", flow:"Sunrise", user:"admin", ...}}, docUri:"/jobs/cf09fd76-b7ad-4273-9428-f9068b374b0a.json", permissions:"xdmp.defaultPermissions().concat([xdmp.permission('flow-develope...", ...}, {database:"6814836408412475919", commit:"auto", update:"true", ...}) -- Range index error: dateTime fn:doc('/jobs/cf09fd76-b7ad-4273-9428-f9068b374b0a.json')/job/text('timeEnded'): XDMP-LEXVAL: Invalid lexical value "N/A"
2019-12-16 20:15:19.028 Debug: IndexerEnv::putRangeIndex: XDMP-RANGEINDEX: xdmp:eval("xdmp.documentInsert(docUri, content, {permissions: xdmp.defaultP...", {content:{job:{jobId:"cf09fd76-b7ad-4273-9428-f9068b374b0a", flow:"Sunrise", user:"admin", ...}}, docUri:"/jobs/cf09fd76-b7ad-4273-9428-f9068b374b0a.json", permissions:"xdmp.defaultPermissions().concat([xdmp.permission('flow-develope...", ...}, {database:"6814836408412475919", commit:"auto", update:"true", ...}) -- Range index error: dateTime fn:doc('/jobs/cf09fd76-b7ad-4273-9428-f9068b374b0a.json')/job/text('timeEnded'): XDMP-LEXVAL: Invalid lexical value "N/A"
2019-12-16 20:15:19.683 Debug: IndexerEnv::putRangeIndex: XDMP-RANGEINDEX: xdmp:eval("xdmp.documentInsert(docUri, content, {permissions: xdmp.defaultP...", {content:{batch:{jobId:"cf09fd76-b7ad-4273-9428-f9068b374b0a", batchId:"90c189b6-92aa-4e0f-bf34-e6882e9c97f4", step:{name:"mapAuto", description:"map auto insurance customer data to the Customer entity definiti...", options:{additionalCollections:[], sourceQuery:"cts.collectionQuery([\"loadAuto\"])", mapping:{name:"Sunrise-mapAuto", version:0}, ...}, ...}, ...}}, docUri:"/jobs/batches/90c189b6-92aa-4e0f-bf34-e6882e9c97f4.json", permissions:"xdmp.defaultPermissions().concat([xdmp.permission('flow-develope...", ...}, {database:"6814836408412475919", commit:"auto", update:"true", ...}) -- Range index error: dateTime fn:doc('/jobs/batches/90c189b6-92aa-4e0f-bf34-e6882e9c97f4.json')/batch/text('timeEnded'): XDMP-LEXVAL: Invalid lexical value "N/A"
2019-12-16 20:15:19.714 Debug: IndexerEnv::putRangeIndex: XDMP-RANGEINDEX: xdmp:eval("xdmp.documentInsert(docUri, content, {permissions: xdmp.defaultP...", {content:{batch:{jobId:"cf09fd76-b7ad-4273-9428-f9068b374b0a", batchId:"6c3153ab-88b8-4a4c-9d2e-7b2ce614c68b", step:{name:"mapAuto", description:"map auto insurance customer data to the Customer entity definiti...", options:{additionalCollections:[], sourceQuery:"cts.collectionQuery([\"loadAuto\"])", mapping:{name:"Sunrise-mapAuto", version:0}, ...}, ...}, ...}}, docUri:"/jobs/batches/6c3153ab-88b8-4a4c-9d2e-7b2ce614c68b.json", permissions:"xdmp.defaultPermissions().concat([xdmp.permission('flow-develope...", ...}, {database:"6814836408412475919", commit:"auto", update:"true", ...}) -- Range index error: dateTime fn:doc('/jobs/batches/6c3153ab-88b8-4a4c-9d2e-7b2ce614c68b.json')/batch/text('timeEnded'): XDMP-LEXVAL: Invalid lexical value "N/A"

**Comment by rjrudin** [comment]
The fix seems to be - in jobs.sjs, don't default timeEnded to "N/A", just set it to null or better yet, don't include it at all since we don't have any value to give it.

**Comment by hao1st** [comment]
@venuiyengar As Rob said, we could change default timeEnded to null. However, I was wondering what the root cause is. What steps have you done? Any error in the content database server (final or stage)?

**Comment by rjrudin** [comment]
@venuiyengar The timeEnded index is configured to ignore invalid values. Thus, while I don't think DHF should be setting "N/A" as the value at any point (it just should set it to null), the only way you'd be getting that error is if your jobs database has "reject" for the "timeEnded" index. 

To verify, try running this in qconsole against data-hub-JOBS:

```
declareUpdate();
xdmp.documentInsert("test1.json", {"jobId":"123", "timeEnded":"N/A"})
```

That should succeed, but I'm guessing it will fail in your environment. 

Please check src/main/hub-internal-config/databases/job-database.json to see how the timeEnded index is configured. If it's "reject", please create a new ticket with the version of DHF that you're using, and possibly the version you upgraded from, as we'd need to determine why it's "reject". Otherwise, check for a custom job-database file somewhere - like in ml-config - that is changing the index to "reject" instead of "ignore", and undo that. Otherwise, the index setting must have been manually configured, so undo that.

**Comment by venuiyengar** [comment]
@rjrudin This issue still exist. I checked the jobs database and it says ignore for timeEnded. When I ran this against the jobs database, it is still generating the same error. This is happening with the quick-start jobs database
declareUpdate();
xdmp.documentInsert("test1.json", {"jobId":"123", "timeEnded":"N/A"})

2020-01-03 11:22:59.910 Debug: IndexerEnv::putRangeIndex: XDMP-RANGEINDEX: xdmp:invoke(function bound (), Element("<options xmlns='xdmp:eval'><database>6814836408412475919</database>...</options>")) -- Range index error: dateTime fn:doc('test1.json')/text('timeEnded'): XDMP-LEXVAL: Invalid lexical value "N/A"

![image](https://user-images.githubusercontent.com/10196200/71735165-afffe800-2e1b-11ea-8ed9-e99ab855d7f3.png)
